daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/processors/parsers/BlobLengthParsers.scala

```diff
@@ -54,14 +53,7 @@ trait ByteChunkWriter { self: PrimParser =>
         writeChunk(array, bytesToPut.toInt)
         remainingBitsToGet -= bitsToGet
       } else {
-        val remainingBits =
-          if (dis.remainingBits.isDefined) {
-            val totalBitsRead = nBits - remainingBitsToGet
-            MaybeULong(dis.remainingBits.get + totalBitsRead)
-          } else {
-            MaybeULong.Nope
-          }
-        self.PENotEnoughBits(start, nBits, remainingBits)
+        self.PENotEnoughBits(start, nBits, dis)
```

Comment:
This change could potentially lead to confusion because blobs are read in chunks?

As an example, let's say we have 3 bytes in the data, our blob size if 4 bytes, and the chunk size is 1 byte.

So `nBits` will be 4 * 8 = 32 bits to read, with only 24 bits available. Because we chunk the reads, we will read 8 bits at a time and successfully read all 24 bits. Then we will go to read the last byte since we need 32 total. That doesn't exist so we call `PENotENoughBits` passing in 32 as the bits we needed. The remaining bits calculation will say there are zero bits available, so the error would say something like

> bit position 24, 32 bits needed, 0 available

Which is may be confusing since we needed 32 bits starting at bitPos = 0, not 24.

Really, we probably want to say:

> bit position 0, 32 bits needed, 24 available

But that's hard because by the time we get to this point the bit position has changed because of the chunk reads, and there aren't actually 24 bits available at this new position, so PENotENoughBits can't easily report that.

Maybe instead we change this line to this:

```scala
self.PENotEnoughBits(start, bitsToGet, dis)
```

Which would report something like this:

> bit position 24, 8 bits needed, 0 available

That's maybe a bit confusing if users aren't aware chunking is going on. They might say "my blob length is 32, why does it say only 8 are needed". But maybe that's fine? Maybe we just teach users that blobs are chunked and if there isn't enough data the error will be about the last chunk read that failed and at least these numbers are accurate?

Reply:
It sounds like when something is to be consumed from the input stream that the starting position should be saved before starting?

Reply:
We can do capture the bitPosition0b (and in some cases we do), but the PE that gets created still uses the state a the time of creation. So we would need to reset the bitPosition back to the beginning of the blob prior to calling PENotEnoughBits. But, that might cause additional problems. For example, the BucketingInputStream has a maximum size it will store, after which it starts throwing away older bytes--if we reset back into those discarded bytes we might get other errors. And with a big blob that doesn't seem impossible.

Maybe alternatively we just save the DataLocation instead of the bitPosition from the beginning of the blob and manually create a PE with that DataLocation?

There is an `PENotEnoughBits` function that does accept a DataLocation just below the one you modified:
```
  def PENotEnoughBits(
    pstate: PState,
    sfl: SchemaFileLocation,
    dataLoc: DataLocation,
    neededBits: Long,
    remainingBits: MaybeULong,
  ) = {
```
That gets a correct starting position. And we would need to adjust your `remainingBits` logic to use the `bitPosition` from the `DataLocation` instead of the `DataInputstream` and account for differences between the DataInputStream and DataLocation bitPositions, but I *think* that should be doable? Might also make sense to have the PENotEnoughBits function you modified to call this one,  and move your logic into that so they use the same logic.

All doable, but could get messy.

Reply:
The `currentLocation` member in `PState` has `bitPos1b` but not `bitPos0b`. I ended up storing the value of `bitPos0b` at the beginning of the parse and then using that in the error message if need be. The chunk writer now returns the number of bits actually written instead of handling the error there and the caller creates the error if the expectation isn't met.